src/base/dof_map_constraints.C

```diff
@@ -431,6 +431,9 @@ private:
         std::vector<bool> is_boundary_node(elem->n_nodes(), false),
           is_boundary_edge(elem->n_edges(), false),
           is_boundary_side(elem->n_sides(), false);
+
+        // We also maintain a separate list of nodeset-based boundary nodes
+        std::vector<bool> is_boundary_nodeset(elem->n_nodes(), false);
```

Comment:
I think read-only is still supposed to be safe (in C++11, anyway; the C++03 standard on data races says "what's a thread?"); what the standard allows vector<bool> to screw up is concurrent writes to different elements.  https://stackoverflow.com/questions/16130494/are-stdmap-and-stdvector-thread-safe